### ISSUE TITLE
Printing CRDs with priority set will hide columns 

### ISSUE THREAD

**Issue by soltysh** [kubernetes/kubernetes#65196]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

Create a CRD:
```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: foos.samplecontroller.k8s.io
spec:
  group: samplecontroller.k8s.io
  version: v1alpha1
  names:
    kind: Foo
    plural: foos
  scope: Namespaced
  validation:
    openAPIV3Schema:
      properties:
        spec:
          properties:
            replicas:
              type: integer
              minimum: 1
              maximum: 10
  additionalPrinterColumns:
  - name: replicas
    type: integer
    JSONPath: .spec.replicas
    priority: 1
```
and then its instance:
```yaml
apiVersion: samplecontroller.k8s.io/v1alpha1
kind: Foo
metadata:
  name: example-foo
spec:
  deploymentName: example-foo
  replicas: 1
```

Invoking `kubectl get foos.samplecontroller.k8s.io` will return only the `NAME` column omitting the `additionalPrinterColumns` that have `priority` set to a value > 0. 

Invoking `kubectl get foos.samplecontroller.k8s.io -o wide` will return all the columns properly.

The culprit lies here: https://github.com/kubernetes/kubernetes/blob/23cd1434e69c1a984e8c24c875c19ccbdd0ba2fe/pkg/printers/humanreadable.go#L351 we should probably calculate the number of columns and only after exceeding some limit print those with lower priority.

/cc @sttts

**Comment by smarterclayton** [comment]
This was deliberate.  Priority > 0 is the same as wide.  The door was left open to automatically show columns.  However, keep in mind that users who expect to script interactions with get (grep, etc) or who have varying terminal windows may be annoyed at you for hiding or showing data based on width (inconsistent output is frustrating to users).  So we need to be careful before we get too clever here.

**Comment by smarterclayton** [comment]
In general this problem is a subset of "kubectl doesn't react well to differing terminal widths" - it might make sense to have a couple of high level principles about how we solve that (wrapping, better help layouts, reactive tables) before we make specific changes to just this code.  When we did this in `rhc` it was very expensive to get to "good", so some of kubectl reflects the "good enough" mindset.  Off the top of my head, areas that wrap poorly:

1. help and flag options (human reading our output)
2. get
3. describe and big tables
4. explain output
5. some error messages

**Comment by soltysh** [comment]
Sounds like a reasonable topic for one of our next sig-cli calls. I'll add a note to the agenda, but since tomorrow's already full it'll overflow into the next one on July 4th.

**Comment by liggitt** [comment]
this is working as designed with respect to custom resources

/close
